enhancements/deprecated-bundles.md

```diff
@@ -0,0 +1,117 @@
+approvers:
+  - "@ecordell"
+  - "@shawn-hurley"
+  - "@kevinrizza"
+creation-date: 2020-07-15
+last-updated: 2020-07-15
+status: implementable
+---
+
+# Deprecating Bundles
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+
+## Summary
+
+OPM will support marking versions as deprecated, truncating the update graph of a given package, and appending to the tail of channel. Deprecated versions will not be installable. If a deprecated version is installed prior to upgrade appropriate alerts will be fired on cluster. 
+
+## Motivation
+
+The default set of Operators provided on OCP clusters will soon (4.6+) be delivered through indexes and each OCP minor version will have a unique set of index images. Today, packages in one index will be duplicated to the next minor version index when it is available. This means operator owners releasing through the redhat, certified, or community pipelines will now need a way to specify which versions of their operator are incompatible or unsupported in the next minor version index.
+
+### Terminology
+
+#### Deprecated bundle
+
+A deprecated bundle is an upgradable but uninstallable bundle in the update graph of an operator.
+
+#### Terminated bundle
+
+A terminated bundle is neither upgradable nor installable.
+
+### Goals
+
+- A deprecated bundle in the update graph and any of its replacements (terminated bundles) will not be installed by OLM
+- Operator Owners don’t need to release a new version in order to specify which bundle is deprecated
+- The index add command is enhanced to un-do this termination if needed
+- The deprecation is specified after promotion (not during) and before publishing of the index image
+- Alerts are fired on cluster when a deprecated version is installed
+
+### Non-Goals
+
+- OLM supports upgrading from a terminated bundle
+- Alerts are fired if terminated bundles exist on cluster
+- OLM prevents cluster upgrades based on whether versions are terminated in the upgraded cluster’s index
+
+## Proposal
+
+We propose to add a command to `opm index` called `deprecate` which takes in a set of bundle images, an index, and all the same arguments as `opm index add`. The command modifies the index by removing all versions that come after the specified bundle image in the update graph. In this way, versions that were removed from the index will not receive updates if subscriptions still exist for them on cluster. A new field is added to the operator-registry database table `operatorbundle` called `deprecated` and the bundle images specified in the command will be marked as such.
+
+The operator-registry APIs will be modified to ensure that OLM does not allow install of deprecated bundles. This also means appropriate status reporting in the subscription and installplan need to be added to OLM, and alerts will be fired if a deprecated version is installed.
```

Comment:
Since this targets 4.6, OLM will query for bundles directly and resolve them with the new resolver. 

That means that we probably want a new `property` type, like `olm.deprecated`/`olm.obsolete`/`olm.not-installble`, and then in the resolver we would filter out any options that have this property set.